firefox_ui_harness/runtests.py

```diff
@@ -7,71 +7,57 @@
 
 
-def startTestRunner(runner_class, options, tests):
-    install_folder = None
-
-    try:
-        # Prepare the workspace path so that all temporary data can be stored inside it.
-        if options.workspace_path:
-            path = os.path.expanduser(options.workspace_path)
-            options.workspace = os.path.abspath(path)
-
-            if not os.path.exists(options.workspace):
-                os.makedirs(options.workspace)
-        else:
-            options.workspace = tempfile.mkdtemp('.{}'.format(os.path.basename(sys.argv[0])))
-
-        options.logger.info('Using workspace for temporary data: "{}"'.format(options.workspace))
+class FirefoxUIHarness(MarionetteHarness):
+    def __init__(self,
+                 runner_class=FirefoxUITestRunner,
+                 parser_class=FirefoxUIArguments):
+        MarionetteHarness.__init__(self, runner_class, parser_class)
+        self.install_folder = None
 
+    def process_args(self):
+        # TODO move installer to mozharness script.
         # If the specified binary is an installer it needs to be installed
-        if options.installer:
-            installer = os.path.realpath(options.installer)
-
-            dest_folder = os.path.join(options.workspace, 'binary')
-            options.logger.info('Installing application "%s" to "%s"' % (installer,
-                                                                         dest_folder))
-            install_folder = mozinstall.install(installer, dest_folder)
-            options.binary = mozinstall.get_binary(install_folder, 'firefox')
-
-        runner = runner_class(**vars(options))
-        runner.run_tests(tests)
-
-    finally:
-        # Ensure to uninstall the binary if it has been installed before
-        if install_folder and os.path.exists(install_folder):
-            options.logger.info('Uninstalling application at "%s"' % install_folder)
-            mozinstall.uninstall(install_folder)
-
-    return runner
-
-
-def cli(runner_class=FirefoxUITestRunner, parser_class=FirefoxUIArguments):
-    parser = parser_class(usage='%(prog)s [options] test_file_or_dir <test_file_or_dir> ...')
-    structured.commandline.add_logging_group(parser)
-    args = parser.parse_args()
-    parser.verify_usage(args)
-
-    logger = structured.commandline.setup_logging(
-        args.logger_name, args, {'mach': sys.stdout})
-    args.logger = logger
-
-    try:
-        runner = startTestRunner(runner_class, args, args.tests)
-        if runner.failed > 0:
-            sys.exit(10)
-
-    except Exception:
-        logger.error('Failure during execution of the update test.',
-                     exc_info=True)
-        sys.exit(1)
-
+        if self.args.installer:
+            installer = os.path.realpath(self.args.installer)
+
+            if not self.args.workspace:
+                self.args.installer_workspace = tempfile.mkdtemp(
+                    '.{}'.format(os.path.basename(sys.argv[0]))
```

Comment:
Ok, so while talking about firefox-ui-tests we should not mix up any decisions about what gets done in mozmill-ci with workspaces. This is just about the plain `firefox-ui-tests` and `firefox-ui-update` CLI commands. For those I think we should have a behavior which is as close as possible to Marionette:

* **Profiles**: If no `--workspace` option is specified for our tests, no profile should remain at the end. If the option has been given then the profile will exist for examination. The only problem here is for our update tests which internally run Marionette twice and as a result two profiles will remain. Not sure at the moment if we should keep that or if the formerly profile has to be deleted. I assume this is even a question for Marionette and not only us given that this feature of multiple runs is in Marionette proper now.

* **Installer**: If no `--workspace``option is specified the binary gets installed under TMP and is also getting uninstalled. With the option set, the binary gets installed into the workspace AND also gets uninstalled. The latter is important especially for Windows given that the uninstaller cleans-up the registry and such.

If we cannot get those behaviors with a virtual internal workspace setting if it hasn't been specified via the command line, we should completely get rid of it and store everything under individual subfolder of TMP.